File: apps/mobile/src/core/OverseerrService.ts
/**
 * Overseerr API Service for requesting media
 * Requires user to enable and provide URL + API key in settings
 */

import { isOverseerrEnabled, getOverseerrUrl, getOverseerrApiKey } from './SettingsData';

const CACHE_TTL = 5 * 60 * 1000; // 5 minutes

// Status enums from Overseerr API
export const MediaRequestStatus = {
  PENDING: 1,
  APPROVED: 2,
  DECLINED: 3,
} as const;

export const MediaInfoStatus = {
  UNKNOWN: 1,
  PENDING: 2,
  PROCESSING: 3,
  PARTIALLY_AVAILABLE: 4,
  AVAILABLE: 5,
} as const;

export type OverseerrStatus =
  | 'not_requested'
  | 'pending'
  | 'approved'
  | 'declined'
  | 'processing'
  | 'partially_available'
  | 'available'
  | 'unknown';

export interface OverseerrMediaStatus {
  status: OverseerrStatus;
  requestId?: number;
  canRequest: boolean;
}

export interface OverseerrRequestResult {
  success: boolean;
  requestId?: number;
  status?: OverseerrStatus;
  error?: string;
}

interface OverseerrUser {
  id: number;
  email: string;
  username: string;
  permissions: number;
}

interface MediaRequest {
  id: number;
  status: number;
  media: {
    id: number;
    tmdbId: number;
    mediaType: string;
    status: number;
  };
}

interface MediaInfo {
  id: number;
  tmdbId: number;
  status: number;
  requests?: MediaRequest[];
}

interface MovieDetails {
  id: number;
  mediaInfo?: MediaInfo;
}

interface TvDetails {
  id: number;
  mediaInfo?: MediaInfo;
}

// In-memory cache for media status
const statusCache = new Map<string, { status: OverseerrMediaStatus; timestamp: number }>();

/**
 * Normalize Overseerr URL - ensure it has /api/v1 suffix and no trailing slash
 */
function normalizeUrl(url: string): string {
  let normalized = url.trim();
  // Remove trailing slash
  if (normalized.endsWith('/')) {
    normalized = normalized.slice(0, -1);
  }
  // Don't add /api/v1 here - we'll add it per request
  return normalized;
}

/**
 * Make authenticated request to Overseerr API
 */
async function overseerrFetch<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const baseUrl = getOverseerrUrl();
  const apiKey = getOverseerrApiKey();

  if (!baseUrl || !apiKey) {
    throw new Error('Overseerr not configured');
  }

  const url = `${normalizeUrl(baseUrl)}/api/v1${endpoint}`;

  const response = await fetch(url, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      'X-Api-Key': apiKey,
      ...options.headers,
    },
  });

  if (!response.ok) {
    const errorText = await response.text().catch(() => 'Unknown error');
    throw new Error(`Overseerr API error (${response.status}): ${errorText}`);
  }

  return response.json();
}

/**
 * Check if Overseerr is enabled and has URL + API key configured
 */
export function isOverseerrReady(): boolean {
  return isOverseerrEnabled() && !!getOverseerrUrl() && !!getOverseerrApiKey();
}

/**
 * Validate Overseerr connection with provided credentials
 */
export async function validateOverseerrConnection(
  url: string,
  apiKey: string
): Promise<{ valid: boolean; username?: string; error?: string }> {
  try {
    const normalizedUrl = normalizeUrl(url);
    const response = await fetch(`${normalizedUrl}/api/v1/auth/me`, {
      headers: {
        'Content-Type': 'application/json',
        'X-Api-Key': apiKey,
      },
    });

    if (!response.ok) {
      if (response.status === 401 || response.status === 403) {
        return { valid: false, error: 'Invalid API key' };
      }
      return { valid: false, error: `Server error (${response.status})` };
    }

    const user: OverseerrUser = await response.json();
    return { valid: true, username: user.username || user.email };
  } catch (error) {
    console.log('[OverseerrService] Connection validation error:', error);
    if (error instanceof TypeError && error.message.includes('Network')) {
      return { valid: false, error: 'Unable to connect to server' };
    }
    return { valid: false, error: 'Connection failed' };
  }
}

/**
 * Convert API status codes to human-readable status
 */
function parseMediaStatus(mediaInfo?: MediaInfo): OverseerrMediaStatus {
  if (!mediaInfo) {
    return { status: 'not_requested', canRequest: true };
  }

  // Check media availability status first
  switch (mediaInfo.status) {
    case MediaInfoStatus.AVAILABLE:
      return { status: 'available', canRequest: false };
    case MediaInfoStatus.PARTIALLY_AVAILABLE:
      return { status: 'partially_available', canRequest: true };
    case MediaInfoStatus.PROCESSING:
      return { status: 'processing', canRequest: false };
  }

  // Check request status if media not available
  const latestRequest = mediaInfo.requests?.[0];
  if (latestRequest) {
    switch (latestRequest.status) {
      case MediaRequestStatus.PENDING:
        return { status: 'pending', requestId: latestRequest.id, canRequest: false };
      case MediaRequestStatus.APPROVED:
        return { status: 'approved', requestId: latestRequest.id, canRequest: false };
      case MediaRequestStatus.DECLINED:
        return { status: 'declined', requestId: latestRequest.id, canRequest: true };
    }
  }

  // Default to not requested
  if (mediaInfo.status === MediaInfoStatus.PENDING) {
    return { status: 'pending', canRequest: false };
  }

  return { status: 'not_requested', canRequest: true };
}

/**
 * Get media request status from Overseerr
 */
export async function getMediaStatus(
  tmdbId: number,
  mediaType: 'movie' | 'tv'
): Promise<OverseerrMediaStatus> {
  if (!isOverseerrReady()) {
    return { status: 'unknown', canRequest: false };
  }

  // Check cache
  const cacheKey = `${mediaType}:${tmdbId}`;
  const cached = statusCache.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    console.log(`[OverseerrService] Cache hit for ${cacheKey}`);
    return cached.status;
  }

  try {
    console.log(`[OverseerrService] Fetching status for ${mediaType}:${tmdbId}`);

    const endpoint = mediaType === 'movie' ? `/movie/${tmdbId}` : `/tv/${tmdbId}`;
    const details = await overseerrFetch<MovieDetails | TvDetails>(endpoint);

    const status = parseMediaStatus(details.mediaInfo);

    // Cache the result
    statusCache.set(cacheKey, { status, timestamp: Date.now() });

    console.log(`[OverseerrService] Status for ${cacheKey}:`, status);
    return status;
  } catch (error) {
    console.log('[OverseerrService] Error fetching status:', error);
    // Return unknown but allow request attempt
    return { status: 'unknown', canRequest: true };
  }
}

/**
 * Get available seasons for a TV show from Overseerr
 */
async function getTvSeasons(tmdbId: number): Promise<number[]> {
  try {
    const details = await overseerrFetch<TvDetails & { seasons?: Array<{ seasonNumber: number }> }>(
      `/tv/${tmdbId}`
    );
    // Filter out season 0 (specials) and return season numbers
    return (details.seasons || [])
      .map(s => s.seasonNumber)
      .filter(n => n > 0);
  } catch (error) {
    console.log('[OverseerrService] Error fetching TV seasons:', error);
    return [];
  }
}

/**
 * Request media through Overseerr
 */
export async function requestMedia(
  tmdbId: number,
  mediaType: 'movie' | 'tv',
  is4k: boolean = false
): Promise<OverseerrRequestResult> {
  if (!isOverseerrReady()) {
    return { success: false, error: 'Overseerr not configured' };
  }

  try {
    console.log(`[OverseerrService] Requesting ${mediaType}:${tmdbId} (4K: ${is4k})`);

    // Build request body
    const requestBody: {
      mediaType: string;
      mediaId: number;
      is4k?: boolean;
      seasons?: number[];
    } = {
      mediaType,
      mediaId: tmdbId,
    };

    // Only add is4k if true
    if (is4k) {
      requestBody.is4k = true;
    }

    // For TV shows, we need to specify which seasons to request
    if (mediaType === 'tv') {
      const seasons = await getTvSeasons(tmdbId);
      if (seasons.length === 0) {
        return { success: false, error: 'Could not determine available seasons' };
      }
      requestBody.seasons = seasons;
      console.log(`[OverseerrService] Requesting seasons:`, seasons);
    }

    const response = await overseerrFetch<MediaRequest>('/request', {
      method: 'POST',
      body: JSON.stringify(requestBody),
    });

    // Clear cache for this item
    const cacheKey = `${mediaType}:${tmdbId}`;
    statusCache.delete(cacheKey);

    console.log('[OverseerrService] Request created:', response);

    // Determine status from response
    let status: OverseerrStatus = 'pending';
    if (response.status === MediaRequestStatus.APPROVED) {
      status = 'approved';
    }

    return {
      success: true,
      requestId: response.id,
      status,
    };
  } catch (error) {
    console.log('[OverseerrService] Error creating request:', error);
    const message = error instanceof Error ? error.message : 'Request failed';
    return { success: false, error: message };
  }
}

/**
 * Clear the status cache
 */
export function clearOverseerrCache(): void {
  statusCache.clear();
  console.log('[OverseerrService] Cache cleared');
}

/**
 * Clear cache for a specific item
 */
export function clearOverseerrCacheItem(tmdbId: number, mediaType: 'movie' | 'tv'): void {
  const cacheKey = `${mediaType}:${tmdbId}`;
  statusCache.delete(cacheKey);
  console.log(`[OverseerrService] Cache cleared for ${cacheKey}`);
}

File: apps/macos/FlixorMac/Services/Overseerr/OverseerrService.swift
//
//  OverseerrService.swift
//  FlixorMac
//
//  Overseerr API service for requesting media
//  Requires user to enable and provide URL + API key in settings
//

import Foundation

@MainActor
class OverseerrService: ObservableObject {
    // MARK: - Singleton

    static let shared = OverseerrService()

    // MARK: - Configuration

    private let cacheTTL: TimeInterval = 5 * 60 // 5 minutes

    // MARK: - Cache

    private var cache: [String: (status: OverseerrMediaStatus, timestamp: Date)] = [:]

    // MARK: - Computed Properties

    private var isEnabled: Bool {
        UserDefaults.standard.bool(forKey: "overseerrEnabled")
    }

    private var serverUrl: String? {
        let url = UserDefaults.standard.string(forKey: "overseerrUrl")
        return url?.isEmpty == false ? url : nil
    }

    private var apiKey: String? {
        let key = UserDefaults.standard.string(forKey: "overseerrApiKey")
        return key?.isEmpty == false ? key : nil
    }

    // MARK: - Initialization

    private init() {}

    // MARK: - Public API

    /// Check if Overseerr is enabled and has URL + API key configured
    func isReady() -> Bool {
        return isEnabled && serverUrl != nil && apiKey != nil
    }

    /// Validate Overseerr connection with provided credentials
    func validateConnection(url: String, apiKey: String) async -> OverseerrConnectionResult {
        do {
            let normalizedUrl = normalizeUrl(url)
            guard let requestUrl = URL(string: "\(normalizedUrl)/api/v1/auth/me") else {
                return OverseerrConnectionResult(valid: false, username: nil, error: "Invalid URL")
            }

            var request = URLRequest(url: requestUrl)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")

            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                return OverseerrConnectionResult(valid: false, username: nil, error: "Invalid response")
            }

            if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                return OverseerrConnectionResult(valid: false, username: nil, error: "Invalid API key")
            }

            if httpResponse.statusCode != 200 {
                return OverseerrConnectionResult(valid: false, username: nil, error: "Server error (\(httpResponse.statusCode))")
            }

            let user = try JSONDecoder().decode(OverseerrUser.self, from: data)
            let username = user.username ?? user.email ?? "user"
            return OverseerrConnectionResult(valid: true, username: username, error: nil)
        } catch {
            print("[OverseerrService] Connection validation error: \(error)")
            return OverseerrConnectionResult(valid: false, username: nil, error: "Connection failed")
        }
    }

    /// Get media request status from Overseerr
    func getMediaStatus(tmdbId: Int, mediaType: String) async -> OverseerrMediaStatus {
        guard isReady() else {
            return .notConfigured
        }

        // Check cache
        let cacheKey = "\(mediaType):\(tmdbId)"
        if let cached = cache[cacheKey],
           Date().timeIntervalSince(cached.timestamp) < cacheTTL {
            print("[OverseerrService] Cache hit for \(cacheKey)")
            return cached.status
        }

        do {
            print("[OverseerrService] Fetching status for \(mediaType):\(tmdbId)")

            let endpoint = mediaType == "movie" ? "/movie/\(tmdbId)" : "/tv/\(tmdbId)"
            let data = try await makeRequest(endpoint: endpoint)

            let status: OverseerrMediaStatus
            if mediaType == "movie" {
                let details = try JSONDecoder().decode(OverseerrMovieDetails.self, from: data)
                status = parseMediaStatus(mediaInfo: details.mediaInfo)
            } else {
                let details = try JSONDecoder().decode(OverseerrTvDetails.self, from: data)
                status = parseMediaStatus(mediaInfo: details.mediaInfo)
            }

            // Cache the result
            cache[cacheKey] = (status: status, timestamp: Date())

            print("[OverseerrService] Status for \(cacheKey): \(status.status.rawValue)")
            return status
        } catch {
            print("[OverseerrService] Error fetching status: \(error)")
            return OverseerrMediaStatus(status: .unknown, canRequest: true)
        }
    }

    /// Request media through Overseerr
    func requestMedia(tmdbId: Int, mediaType: String, is4k: Bool = false) async -> OverseerrRequestResult {
        guard isReady() else {
            return OverseerrRequestResult(success: false, requestId: nil, status: nil, error: "Overseerr not configured")
        }

        do {
            print("[OverseerrService] Requesting \(mediaType):\(tmdbId) (4K: \(is4k))")

            // Build request body
            var requestBody: [String: Any] = [
                "mediaType": mediaType,
                "mediaId": tmdbId
            ]

            if is4k {
                requestBody["is4k"] = true
            }

            // For TV shows, we need to specify which seasons to request
            if mediaType == "tv" {
                let seasons = await getTvSeasons(tmdbId: tmdbId)
                if seasons.isEmpty {
                    return OverseerrRequestResult(success: false, requestId: nil, status: nil, error: "Could not determine available seasons")
                }
                requestBody["seasons"] = seasons
                print("[OverseerrService] Requesting seasons: \(seasons)")
            }

            let bodyData = try JSONSerialization.data(withJSONObject: requestBody)
            let data = try await makeRequest(endpoint: "/request", method: "POST", body: bodyData)

            // Clear cache for this item
            let cacheKey = "\(mediaType):\(tmdbId)"
            cache.removeValue(forKey: cacheKey)

            let response = try JSONDecoder().decode(OverseerrMediaRequest.self, from: data)
            print("[OverseerrService] Request created: \(response.id)")

            // Determine status from response
            var status: OverseerrStatus = .pending
            if response.status == MediaRequestStatusCode.approved.rawValue {
                status = .approved
            }

            return OverseerrRequestResult(success: true, requestId: response.id, status: status, error: nil)
        } catch {
            print("[OverseerrService] Error creating request: \(error)")
            return OverseerrRequestResult(success: false, requestId: nil, status: nil, error: error.localizedDescription)
        }
    }

    /// Clear the status cache
    func clearCache() {
        cache.removeAll()
        print("[OverseerrService] Cache cleared")
    }

    /// Clear cache for a specific item
    func clearCacheItem(tmdbId: Int, mediaType: String) {
        let cacheKey = "\(mediaType):\(tmdbId)"
        cache.removeValue(forKey: cacheKey)
        print("[OverseerrService] Cache cleared for \(cacheKey)")
    }

    // MARK: - Private Helpers

    private func normalizeUrl(_ url: String) -> String {
        var normalized = url.trimmingCharacters(in: .whitespacesAndNewlines)
        if normalized.hasSuffix("/") {
            normalized = String(normalized.dropLast())
        }
        return normalized
    }

    private func makeRequest(endpoint: String, method: String = "GET", body: Data? = nil) async throws -> Data {
        guard let serverUrl = serverUrl, let apiKey = apiKey else {
            throw NSError(domain: "OverseerrService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Overseerr not configured"])
        }

        let normalizedUrl = normalizeUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/api/v1\(endpoint)") else {
            throw NSError(domain: "OverseerrService", code: -2, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")
        request.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "OverseerrService", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            let errorText = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw NSError(domain: "OverseerrService", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "Overseerr API error (\(httpResponse.statusCode)): \(errorText)"])
        }

        return data
    }

    private func parseMediaStatus(mediaInfo: OverseerrMediaInfo?) -> OverseerrMediaStatus {
        guard let mediaInfo = mediaInfo else {
            return OverseerrMediaStatus(status: .notRequested, canRequest: true)
        }

        // Check media availability status first
        switch mediaInfo.status {
        case MediaInfoStatusCode.available.rawValue:
            return OverseerrMediaStatus(status: .available, canRequest: false)
        case MediaInfoStatusCode.partiallyAvailable.rawValue:
            return OverseerrMediaStatus(status: .partiallyAvailable, canRequest: true)
        case MediaInfoStatusCode.processing.rawValue:
            return OverseerrMediaStatus(status: .processing, canRequest: false)
        default:
            break
        }

        // Check request status if media not available
        if let latestRequest = mediaInfo.requests?.first {
            switch latestRequest.status {
            case MediaRequestStatusCode.pending.rawValue:
                return OverseerrMediaStatus(status: .pending, requestId: latestRequest.id, canRequest: false)
            case MediaRequestStatusCode.approved.rawValue:
                return OverseerrMediaStatus(status: .approved, requestId: latestRequest.id, canRequest: false)
            case MediaRequestStatusCode.declined.rawValue:
                return OverseerrMediaStatus(status: .declined, requestId: latestRequest.id, canRequest: true)
            default:
                break
            }
        }

        // Default to not requested
        if mediaInfo.status == MediaInfoStatusCode.pending.rawValue {
            return OverseerrMediaStatus(status: .pending, canRequest: false)
        }

        return OverseerrMediaStatus(status: .notRequested, canRequest: true)
    }

    private func getTvSeasons(tmdbId: Int) async -> [Int] {
        do {
            let data = try await makeRequest(endpoint: "/tv/\(tmdbId)")
            let details = try JSONDecoder().decode(OverseerrTvDetails.self, from: data)
            // Filter out season 0 (specials) and return season numbers
            return (details.seasons ?? [])
                .map { $0.seasonNumber }
                .filter { $0 > 0 }
        } catch {
            print("[OverseerrService] Error fetching TV seasons: \(error)")
            return []
        }
    }
}
